packages/typescript-estree/src/convert.ts

```diff
@@ -1695,6 +1695,7 @@ export class Converter {
           object: this.convertChild(node.expression),
           property: this.convertChild(node.name),
           computed: false,
+          optionalChain: node.questionDotToken !== undefined,
```

Comment:
Neither Acorn nor Esprima seem to have something for optional chaining, but `babel-eslint9`'s AST uses different nodes for each portion of the entire OptionalExpression, with `optional: true` if the `?.` token was used or `optional: false` if it is not.

If reading left-to-right (from the deepest AST node to the shallowest), regular `MemberExpression` / `CallExpression` nodes will be emitted until the first `?.` is encountered, at which point everything after it is either an `OptionalMemberExpression` or `OptionalCallExpression`, with `optional: true` or `optional: false` as needed. This is a consequence of how short-circuiting was coded in the spec itself, and is probably useful for eslint plugins as well.

[*]: Actually, `OptionalCallExpression` is missing `optional: false` entirely when it's not optional (effectively being `optional: undefined`), possibly a bug.

Reply:
Interesting. For some reason I didn't think about checking babel's AST in this case.
TypeScript just does a very, very simple transform - it just adds the `questionDotToken` prop to the member expression.
I'll go back and revisit this then.

Dumping the babel ASTs here for reference:

<details>
<summary> <pre>a.b.c.d.e?.f</pre> </summary>

```json
{
  "type": "Program",
  "comments": [],
  "sourceType": "module",
  "body": [
    {
      "type": "ExpressionStatement",
      "expression": {
        "type": "OptionalMemberExpression",
        "object": {
          "type": "MemberExpression",
          "object": {
            "type": "MemberExpression",
            "object": {
              "type": "MemberExpression",
              "object": {
                "type": "MemberExpression",
                "object": {
                  "type": "Identifier",
                  "name": "a"
                },
                "property": {
                  "type": "Identifier",
                  "name": "b"
                },
                "computed": false
              },
              "property": {
                "type": "Identifier",
                "name": "c"
              },
              "computed": false
            },
            "property": {
              "type": "Identifier",
              "name": "d"
            },
            "computed": false
          },
          "property": {
            "type": "Identifier",
            "name": "e"
          },
          "computed": false
        },
        "property": {
          "type": "Identifier",
          "name": "f"
        },
        "computed": false,
        "optional": true
      }
    }
  ]
}
```

</details>


<details>
<summary> <pre>a.b.c.d.e?.f()</pre> </summary>

```json
{
  "type": "Program",
  "comments": [],
  "sourceType": "module",
  "body": [
    {
      "type": "ExpressionStatement",
      "expression": {
        "type": "OptionalCallExpression",
        "callee": {
          "type": "OptionalMemberExpression",
          "object": {
            "type": "MemberExpression",
            "object": {
              "type": "MemberExpression",
              "object": {
                "type": "MemberExpression",
                "object": {
                  "type": "MemberExpression",
                  "object": {
                    "type": "Identifier",
                    "name": "a"
                  },
                  "property": {
                    "type": "Identifier",
                    "name": "b"
                  },
                  "computed": false
                },
                "property": {
                  "type": "Identifier",
                  "name": "c"
                },
                "computed": false
              },
              "property": {
                "type": "Identifier",
                "name": "d"
              },
              "computed": false
            },
            "property": {
              "type": "Identifier",
              "name": "e"
            },
            "computed": false
          },
          "property": {
            "type": "Identifier",
            "name": "f"
          },
          "computed": false,
          "optional": true
        },
        "arguments": []
      }
    }
  ]
}
```

</details>

Reply:
One detail that was pointed out to me in a conversation -- parens break the propagation of `Optional{Member,Call}Expression`.

In `(a.b?.c.d).e?.f`, the `.e` part is a regular `MemberExpression` as, if `?.c` short-circuits, the parens would cause `(undefined).e` to be evaluated.

Maybe TypeScript could get away with _only_ adding a `questionDotToken` if they have other means of keeping track of parens; but babel doesn't.